sale_blanket_order/models/blanket_orders.py

```diff
@@ -0,0 +1,719 @@
+
+        if self.partner_id.user_id:
+            values["user_id"] = self.partner_id.user_id.id
+        if self.partner_id.team_id:
+            values["team_id"] = self.partner_id.team_id.id
+        self.update(values)
+
+    def unlink(self):
+        for order in self:
+            if order.state not in ("draft", "expired") or order._check_active_orders():
+                raise UserError(
+                    _(
+                        "You can not delete an open blanket or "
+                        "with active sale orders! "
+                        "Try to cancel it before."
+                    )
+                )
+        return super().unlink()
+
+    def _validate(self):
+        try:
+            today = fields.Date.today()
+            for order in self:
+                assert order.validity_date, _("Validity date is mandatory")
+                assert order.validity_date > today, _(
+                    "Validity date must be in the future"
+                )
+                assert order.partner_id, _("Partner is mandatory")
+                assert len(order.line_ids) > 0, _("Must have some lines")
+                order.line_ids._validate()
+        except AssertionError as e:
+            raise UserError(e) from e
+
+    def set_to_draft(self):
+        for order in self:
+            order.write({"state": "draft", "confirmed": False})
+        return True
+
+    def action_confirm(self):
+        self._validate()
+        for order in self:
+            sequence_obj = self.env["ir.sequence"]
+            if order.company_id:
+                sequence_obj = sequence_obj.with_company(order.company_id.id)
+            name = sequence_obj.next_by_code("sale.blanket.order")
+            order.write({"confirmed": True, "name": name})
+        return True
+
+    def _check_active_orders(self):
+        for order in self.filtered("sale_count"):
+            for so in order._get_sale_orders():
+                if so.state not in ("cancel"):
+                    return True
+        return False
+
+    def action_cancel(self):
+        for order in self:
+            if order._check_active_orders():
+                raise UserError(
+                    _(
+                        "You can not delete a blanket order with opened "
+                        "sale orders! "
+                        "Try to cancel them before."
+                    )
+                )
+            order.write({"state": "expired"})
+        return True
+
+    def action_view_sale_orders(self):
+        sale_orders = self._get_sale_orders()
+        action = self.env.ref("sale.action_orders").read()[0]
+        if len(sale_orders) > 0:
+            action["domain"] = [("id", "in", sale_orders.ids)]
+            action["context"] = [("id", "in", sale_orders.ids)]
+        else:
+            action = {"type": "ir.actions.act_window_close"}
+        return action
+
+    def action_view_sale_blanket_order_line(self):
+        action = self.env.ref(
```

Comment:
@nguyenminhchien, to prevent permission error
```suggestion
        action = self.env["ir.actions.act_window"]._for_xml_id(
            "sale_blanket_order.act_open_sale_blanket_order_lines_view_tree")
```

Reply:
thanks, i have updated them